### PR TITLE
Filter out discrete sites

### PR DESCRIPTION
**Pull request by fehiepsi**
This filters out discrete sites in some utilities. In addition, I correct the shape of `log_det` term in potential_energy computation, which is useful for enum potential energy.

### CODE REVIEW

File: numpyro/infer/util.py [hunk 1]
```diff
@@ -120,6 +121,21 @@ def substitute_fn(site):
             (return_deterministic and (v['type'] == 'deterministic'))}
 
 
+def _unconstrain_reparam(params, site):
```
Comment (fehiepsi) [inline numpyro/infer/util.py:124]: I split this out to use it for `enum_potential_energy`.

File: numpyro/infer/util.py [hunk 2]
```diff
@@ -120,6 +121,21 @@ def substitute_fn(site):
...
+    name = site['name']
+    if name in params:
+        p = params[name]
+        t = biject_to(site['fn'].support)
+        value = t(p)
+
+        log_det = t.log_abs_det_jacobian(p, value)
+        log_det = sum_rightmost(log_det, np.ndim(log_det) - np.ndim(value) + len(site['fn'].event_shape))
```
Comment (fehiepsi) [inline numpyro/infer/util.py:132]: @eb8680 will this resolve your logic at [this line](https://github.com/pyro-ppl/numpyro/pull/572/files#diff-fee397191e0cc75a2e33eea8f7ecb6d5R197)? Here, `log_det` shape and `site.log_prob` shape will have the same shape (or broadcastable).

Comment (eb8680) [reply numpyro/infer/util.py:132]: I think that logic was to account for dimensions that weren't marked as plate or enumeration dimensions, which includes both event dimensions and batch dimensions that should have been plates but weren't.  My code wasn't tested directly so if your code ensures that all such dimensions have been removed then it's probably an improvement.

Comment (fehiepsi) [reply numpyro/infer/util.py:132]: This only guarantees that this new `log_det` site will have the same `dim_to_name`, which is specified [here](https://github.com/pyro-ppl/numpyro/pull/572/files#diff-bd74ce0e030e5b37b3cb5308e7963720R527), as the corresponding sample site. To be honest, I think this is correct but not sure if it is equivalent to the way you resolved in `enum_potential_energy`. Here this log_det site will be an observed site with event_shape (0,) (it is [unit distribution](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/unit.py)) and batch_shape = log_prob.shape.

**Comment by fehiepsi** [comment]
Thanks for reviewing!

**Pull request merged**